cpp/src/ts/batched_kalman.cu

```diff
@@ -0,0 +1,623 @@
+    //   P0 = np.reshape(invImTT @ (R_bi @ R_bi.T).ravel(), (r, r), order="F")
+    BatchedMatrix I_m_TxT =
+      BatchedMatrix::Identity(r * r, num_batches, Zb.pool()) - b_kron(Tb, Tb);
+    BatchedMatrix invI_m_TxT_x_RRTvec = b_solve(I_m_TxT, RRT.vec());
+    BatchedMatrix P0 = invI_m_TxT_x_RRTvec.mat(r, r);
+    P = P0;
+    // auto& stream = std::cout;
+    // stream.precision(16);
+    // MLCommon::myPrintDevVector("P0", P[0], 4*P0.batches(), stream);
+  }
+
+  // init alpha to zero
+  BatchedMatrix alpha(r, 1, num_batches, Zb.pool(), true);
+
+  // init vs, Fs
+  // In batch-major format.
+  double* d_sumlogFs;
+
+  allocate(d_sumlogFs, num_batches);
+
+  CUDA_CHECK(cudaPeekAtLastError());
+
+  // Reference implementation
+  // For it = 1:nobs
+  //  // 1.
+  //   vs[it] = ys[it] - alpha(0,0);
+  //  // 2.
+  //   Fs[it] = P(0,0);
+
+  //   if(Fs[it] < 0) {
+  //     std::cout << "P=" << P << "\n";
+  //     throw std::runtime_error("ERROR: F < 0");
+  //   }
+  //   3.
+  //   MatrixT K = 1.0/Fs[it] * (T * P * Z.transpose());
+  //   4.
+  //   alpha = T*alpha + K*vs[it];
+  //   5.
+  //   MatrixT L = T - K*Z;
+  //   6.
+  //   P = T * P * L.transpose() + R * R.transpose();
+  //   loglikelihood += std::log(Fs[it]);
+  // }
+
+  batched_kalman_loop(d_ys, nobs, Tb, Zb, RRT, P, alpha, r, d_vs, d_Fs,
+                      d_sumlogFs);
+
+  // Finalize loglikelihood
+  // 7. & 8.
+  // double sigma2 = ((vs.array().pow(2.0)).array() / Fs.array()).mean();
+  // double loglike = -.5 * (loglikelihood + nobs * std::log(sigma2));
+  // loglike -= nobs / 2. * (std::log(2 * M_PI) + 1);
+
+  batched_kalman_loglike(d_vs, d_Fs, d_sumlogFs, nobs, num_batches, d_sigma2,
+                         d_loglike);
+}
+
+// Because the kalman filter is typically called many times within the ARIMA
+// `fit()` method, allocations end up very costly. We avoid this by re-using the
+// variables stored in this global `GPUContext` object.
+class GPUContext {
+ public:
+  int m_p = 0;
+  int m_q = 0;
+  double* d_ys = 0;
+  double* d_vs = 0;
+  double* d_Fs = 0;
+  double* d_loglike = 0;
+  double* d_sigma2 = 0;
+  std::shared_ptr<BatchedMatrixMemoryPool> pool = 0;
+
+  // // batched_jones
+  double* d_ar = 0;
+  double* d_Tar = 0;
+  double* d_ma = 0;
+  double* d_Tma = 0;
+
+  GPUContext(int p, int q) : m_p(p), m_q(q) {
+    d_ys = nullptr;
+    d_vs = nullptr;
+    d_Fs = nullptr;
+    d_loglike = nullptr;
+    d_sigma2 = nullptr;
+    pool = nullptr;
+    d_ar = nullptr;
+    d_Tar = nullptr;
+    d_ma = nullptr;
+    d_Tma = nullptr;
+  }
+
+  // Note: Tried to re-use these device vectors, but it caused segfaults, so we ignore them for now.
+  // thrust::device_vector<double> d_Z_b;
+  // thrust::device_vector<double> d_R_b;
+  // thrust::device_vector<double> d_T_b;
+
+  // Only allocates when the pointer is uninitialized.
+  static void allocate_if_zero(double*& ptr, size_t size) {
+    if (ptr == 0) {
+      MLCommon::allocate(ptr, size);
```

Comment:
please use `deviceAllocator` and/or `device_buffer` instead of explicit device memory management.